tools/analyzer_plugin/test/integration/diagnostics/callback_ref_test.dart

```diff
@@ -0,0 +1,281 @@
+import 'package:test/test.dart';
+import 'package:test_reflective_loader/test_reflective_loader.dart';
+
+import '../test_bases/diagnostic_test_base.dart';
+
+void main() {
+  defineReflectiveSuite(() {
+    defineReflectiveTests(CallbackRefDiagnosticFnComponentTest);
+  });
+}
+
+abstract class CallbackRefDiagnosticTest extends DiagnosticTestBase {
+  @override
+  get errorUnderTest => CallbackRefDiagnostic.code;
+
+  @override
+  get fixKindUnderTest => CallbackRefDiagnostic.fixKind;
+
+  static const refVarDeclarations = '''
+    ChildComponent _someCustomRefName;
+    ChildComponent _anotherCustomRefName;''';
+
+  static const refVarDeclarationsFixedSome = '''
+    final _someCustomRefName = useRef<dynamic>();
+    ChildComponent _anotherCustomRefName;''';
+
+  static const refVarDeclarationsFixedTheRest = '''
+    final _someCustomRefName = useRef<dynamic>();
+    final _anotherCustomRefName = useRef<dynamic>();''';
+
+  static const refUsage = '''
+    void foo() {
+      _someCustomRefName.someMethodName();
+      _someCustomRefName?.anotherMethodName();
+      final bar = _someCustomRefName.someGetter;
+
+      _anotherCustomRefName.someMethodName();
+      _anotherCustomRefName?.anotherMethodName();
+      final baz = _anotherCustomRefName.someGetter;
+    }''';
+
+  static const refUsageFixedSome = '''
+    void foo() {
+      _someCustomRefName.current.someMethodName();
+      _someCustomRefName.current?.anotherMethodName();
+      final bar = _someCustomRefName.current.someGetter;
+
+      _anotherCustomRefName.someMethodName();
+      _anotherCustomRefName?.anotherMethodName();
+      final baz = _anotherCustomRefName.someGetter;
+    }''';
+
+  static const refUsageFixedTheRest = '''
+    void foo() {
+      _someCustomRefName.current.someMethodName();
+      _someCustomRefName.current?.anotherMethodName();
+      final bar = _someCustomRefName.current.someGetter;
+
+      _anotherCustomRefName.current.someMethodName();
+      _anotherCustomRefName.current?.anotherMethodName();
+      final baz = _anotherCustomRefName.current.someGetter;
+    }''';
+
+  static const renderReturn = '''
+    return Fragment()(
+      (Child()
+        ..ref = (ref) {
+          _someCustomRefName = ref;
+        })(props.children),
+      (Child()
+        ..id = 'bar'
+        ..ref = ((ref) => _anotherCustomRefName = ref)
+      )('hi'),
+      (Child())('there'),
+    );
+  ''';
+
+  static const renderReturnFixedSome = '''
+    return Fragment()(
+      (Child()
+        ..ref = _someCustomRefName)(props.children),
+      (Child()
+        ..id = 'bar'
+        ..ref = ((ref) => _anotherCustomRefName = ref)
+      )('hi'),
+      (Child())('there'),
+    );
+  ''';
+
+  static const renderReturnFixedTheRest = '''
+    return Fragment()(
+      (Child()
+        ..ref = _someCustomRefName)(props.children),
+      (Child()
+        ..id = 'bar'
+        ..ref = _anotherCustomRefName
```

Comment:
There are cases here for when ref is being assigned to an arrow callback ref, and a custom ref variable.

Can we also add test cases for when the ref is:
1. a callback ref with a block function body
1. a callback ref that's a tearoff?

Reply:
@greglittlefield-wf the two cases that are being tested are a callback ref with a block function body and a callback ref with an arrow function body.

I will add one for a tearoff.

Reply:
Ohh now I see the block body one, sorry about that. Sounds good, thanks!

Reply:
Made the tearoff diagnostic be error only. Implementing a fix for that seemed like its worthy of a follow-up / discussion.